src/commonMain/kotlin/fr/acinq/lightning/payment/OutgoingPaymentHandler.kt

```diff
@@ -344,17 +344,28 @@ class OutgoingPaymentHandler(val nodeParams: NodeParams, val walletParams: Walle
             )
         }
 
-        val minFinalExpiryDelta = request.paymentRequest.minFinalExpiryDelta ?: Channel.MIN_CLTV_EXPIRY_DELTA
-        val finalExpiry = nodeParams.paymentRecipientExpiryParams.computeFinalExpiry(currentBlockHeight, minFinalExpiryDelta)
-        val finalPayload = PaymentOnion.FinalPayload.createSinglePartPayload(request.amount, finalExpiry, request.paymentRequest.paymentSecret, request.paymentRequest.paymentMetadata)
+        when (request.paymentRequest) {
+            is Bolt11Invoice -> {
+                val minFinalExpiryDelta = request.paymentRequest.minFinalExpiryDelta ?: Channel.MIN_CLTV_EXPIRY_DELTA
+                val finalExpiry = nodeParams.paymentRecipientExpiryParams.computeFinalExpiry(currentBlockHeight, minFinalExpiryDelta)
+                val finalPayload = PaymentOnion.FinalPayload.createSinglePartPayload(request.amount, finalExpiry, request.paymentRequest.paymentSecret, request.paymentRequest.paymentMetadata)
+
+                val invoiceFeatures = request.paymentRequest.features
+                val (trampolineAmount, trampolineExpiry, trampolineOnion) = if (invoiceFeatures.hasFeature(Feature.TrampolinePayment) || invoiceFeatures.hasFeature(Feature.ExperimentalTrampolinePayment)) {
+                    OutgoingPaymentPacket.buildPacket(request.paymentHash, trampolineRoute, finalPayload, null)
+                } else {
+                    OutgoingPaymentPacket.buildTrampolineToNonTrampolinePacket(request.paymentRequest, trampolineRoute, finalPayload)
+                }
+                return Triple(trampolineAmount, trampolineExpiry, trampolineOnion.packet)
+            }
+            is Bolt12Invoice -> {
+                val finalExpiry = nodeParams.paymentRecipientExpiryParams.computeFinalExpiry(currentBlockHeight, CltvExpiryDelta(0))
```

Comment:
I don't think we should assume the recipient is buggy. Increasing the CLTV for no good reason has negative consequences. We could add 1 to prepare for a new block appearing right now but anything more doesn't seem useful to me.

Reply:
> I don't think we should assume the recipient is buggy.

It's not about them being buggy, it's that there is inherent delay in receiving blocks, because that's how bitcoin works. It's very common that one node is a few blocks behind because its node is processing new blocks, and that's particularly true of electrum nodes who do some (expensive) indexing on block data. It's a pity to see payments that successfully reached the recipient fail because of this. We should at least use 3 blocks here to prevent that.

Reply:
I meant buggy not in the sense that they lag behind but in the sense that they didn't plan for it by adding enough CLTV delta in the blinded route. If both the recipient and the sender account for being a few blocks behind we add unnecessary CLTV delta, only one should of them should do that and I think it should be the recipient. The problem is that the spec is not very precise on this, that should be clarified.

Reply:
I think you're missing something here. If the recipient requires a final delta of 18, and they receive an HTLC that expires in 17 blocks, they'll reject that payment. This means that even if the recipient included a final delta in their blinded path, if the sender is lagging behind in terms of blockchain height, and doesn't add a few blocks, the payment will be rejected. That's just too dumb and easy to work around by adding a few blocks on the sender side. That's what we've always been doing for non-blinded payments and IIRC we included that in the rationale section of the spec.

Reply:
OK, I thought that's what the final expiry delta was for. But if the recipient has already included their final expiry delta in the route, then the current code is correct, `nodeParams.paymentRecipientExpiryParams.computeFinalExpiry` already takes care of adding a few blocks.
To compare, the code for Bolt 11 is 

```kotlin
val minFinalExpiryDelta = request.paymentRequest.minFinalExpiryDelta ?: Channel.MIN_CLTV_EXPIRY_DELTA
val finalExpiry = nodeParams.paymentRecipientExpiryParams.computeFinalExpiry(currentBlockHeight, minFinalExpiryDelta)
```

Reply:
> OK, I thought that's what the final expiry delta was for.

The final expiry delta is for the case where as a recipient you send `update_fulfill_htlc` but your peer then goes offline. You need enough time to publish your commit tx and HTLC-success transactions, that's why you set a high enough min final expiry delta.

>  then the current code is correct, nodeParams.paymentRecipientExpiryParams.computeFinalExpiry already takes care of adding a few blocks.

I had missed that! Then we should be all good :+1: